core/trie/proof.go

```diff
@@ -175,5 +192,129 @@ func VerifyProof(root *felt.Felt, key *Key, value *felt.Felt, proofs []ProofNode
 			remainingPath.Truncate(proofNode.Edge.Path.Len())
 		}
 	}
+
 	return expectedHash.Equal(value)
 }
+
+// getExpectedhash effectievly gets the value corresponding to the key given proofs
+// https://github.com/eqlabs/pathfinder/blob/main/crates/merkle-tree/src/tree.rs#L2006
+func getExpectedProofValue(root *felt.Felt, proofKey *Key, nbrKey *Key, proofs []ProofNode, hash hashFunc) (*felt.Felt, error) { // Todo: test
+	if proofKey.Len() != 251 || nbrKey.Len() != 251 { //nolint:gomnd
+		return nil, errors.New("keys not the correct length")
+	}
+
+	commonAncestorKey := nbrKey.commonPrefix(*proofKey)
+	expectedHash := root
+	remainingPath := proofKey
+	height := uint8(0)
+	for _, proofNode := range proofs {
+		if !proofNode.Hash(hash).Equal(expectedHash) {
+			return nil, errors.New("proofNode not the expected hash")
+		}
+		if height == commonAncestorKey.len {
+			return proofNode.Hash(hash), nil
+		}
+		switch {
+		case proofNode.Binary != nil:
+			if remainingPath.Test(remainingPath.Len() - 1) {
+				expectedHash = proofNode.Binary.RightHash
+			} else {
+				expectedHash = proofNode.Binary.LeftHash
+			}
+			remainingPath.RemoveLastBit()
+			height++
+		case proofNode.Edge != nil:
+			if !proofNode.Edge.Path.Equal(remainingPath.SubKey(proofNode.Edge.Path.Len())) {
+				return nil, errors.New("error in key path")
+			}
+			expectedHash = proofNode.Edge.Child
+			remainingPath.Truncate(proofNode.Edge.Path.Len())
+			height += proofNode.Edge.Path.Len()
+		}
+
+	}
+	return nil, errors.New("failed to get proof value")
+}
+
+// VerifyRangeProof verifies the range proof for the given range of keys.
+// ref: https://github.com/ethereum/go-ethereum/blob/v1.14.3/trie/proof.go#L484
+// Note: this currently assumes that the inner keys do not contain the min/max key (ie both proofs exist) // Todo
+// The first/last key and value must correspond to the left/right proofs //Todo we currently assume both proofs are provided, as above
+func VerifyRangeProof(root *felt.Felt, keys []*Key, values []*felt.Felt, proofs [2][]ProofNode, hash hashFunc) (bool, error) {
```

Comment:
Note, the `proofs` is a single array, not two. And the `keys` are `*felt.Felt`, meaning its always leaf. Well, I guess if you add a note on that, it does not matter if its a `Key` or `Felt`.

Reply:
We only require two proofs no? Where each proof is a set of nodes from the root to a key.
Eg [geth](https://github.com/ethereum/go-ethereum/blob/master/trie/proof.go#L539) requires two proofs in their implementation. This is also how I understood it from watching Daniels snapsync video